AzureIoTHub.agent.lib.nut

```diff
@@ -624,6 +628,224 @@ class AzureIoTHub {
+                        }
+                    } else {
+                        onCompleted(AZURE_DPS_ERROR_NOT_REGISTERED, body, null);
+                    }
+                } else if (resp.statuscode == AZURE_DPS_NOT_FOUND_CODE) {
+                    onCompleted(AZURE_DPS_ERROR_NOT_REGISTERED, body, null);
+                } else {
+                    onCompleted(resp.statuscode, body, null);
+                }
+            }.bindenv(this);
+
+            request.sendasync(onSent);
+        }
+
+        // -------------------- PRIVATE METHODS -------------------- //
+
+        function _regOperationStatus(operationId, onCompleted) {
+            local url = format(AZURE_DPS_OP_STATUS_ENDPOINT_FMT, AZURE_DPS_GLOBAL_HOST, _scopeId, _regId, operationId, AZURE_DPS_API_VERSION);
+            local request = http.get(url, _headers);
+            local onSent = null;
+
+            onSent = function(resp) {
+                local body = _parseBody(resp.body);
+
+                local retryAfter = AZURE_DPS_DEFAULT_DELAY;
+
+                if (resp.statuscode == AZURE_DPS_OK_CODE || resp.statuscode == AZURE_DPS_ACCEPTED_CODE) {
+                    if (!("status" in body)) {
+                        server.error("Response body doesn't have the \"status\" field");
+                        onCompleted(AZURE_ERROR_GENERAL, body, null);
+                        return;
+                    }
+
+                    if (body["status"] == "assigned") {
+                        if ("registrationState" in body && "assignedHub" in body["registrationState"]) {
+                            onCompleted(0, body, _connectionString(body["registrationState"]["assignedHub"]));
+                        } else {
+                            server.error("Response body doesn't have the \"registrationState.assignedHub\" field");
+                            onCompleted(AZURE_ERROR_GENERAL, body, null);
+                        }
+                        return;
+                    } else if (body["status"] != "assigning") {
+                        onCompleted(AZURE_DPS_ERROR_NOT_REGISTERED, body, null);
+                        return;
+                    }
+                } else if (resp.statuscode != AZURE_DPS_TOO_MANY_REQS_CODE) {
+                    onCompleted(resp.statuscode, body, null);
+                    return;
+                }
+
+                if ("retry-after" in resp.headers) {
+                    retryAfter = resp.headers["retry-after"].tointeger();
+                }
+                request = http.get(url, _headers);
+                imp.wakeup(retryAfter, @() request.sendasync(onSent));
+            }.bindenv(this);
+
+            request.sendasync(onSent);
+        }
+
+        function _connectionString(iotHubAddr) {
+            return format("HostName=%s;DeviceId=%s;SharedAccessKey=%s", iotHubAddr, _regId, _deviceKey);
+        }
+
+        function _parseBody(body) {
+            try {
+                body = http.jsondecode(body);
```

Comment:
I don't like the idea of modifying the argument even though it's a reference. I would introduce `local result` for that instead.

Reply:
Fixed